snuba/redis.py

```diff
@@ -67,35 +65,77 @@ def wrapper() -> RedisClientType:
                     raise
             raise FailedClusterInitization("Init failed")
 
-        return wrapper
+        return cast(RedisInitFunction, wrapper)
 
     return _retry_inner
 
 
 @_retry(max_retries=settings.REDIS_INIT_MAX_RETRIES)
-def _initialize_redis_cluster() -> RedisClientType:
-    if settings.USE_REDIS_CLUSTER:
-        startup_nodes = settings.REDIS_CLUSTER_STARTUP_NODES
+def _initialize_redis_cluster(config: settings.RedisClusterConfig) -> RedisClientType:
+    if config["use_redis_cluster"]:
+        startup_nodes = config["cluster_startup_nodes"]
         if startup_nodes is None:
-            startup_nodes = [{"host": settings.REDIS_HOST, "port": settings.REDIS_PORT}]
+            startup_nodes = [{"host": config["host"], "port": config["port"]}]
         startup_cluster_nodes = [
             ClusterNode(n["host"], n["port"]) for n in startup_nodes
         ]
         return RetryingStrictRedisCluster(
             startup_nodes=startup_cluster_nodes,
             socket_keepalive=True,
-            password=settings.REDIS_PASSWORD,
+            password=config["password"],
             max_connections_per_node=True,
-            reinitialize_steps=settings.REDIS_REINITIALIZE_STEPS,
+            reinitialize_steps=config["reinitialize_steps"],
         )
     else:
         return StrictRedis(
-            host=settings.REDIS_HOST,
-            port=settings.REDIS_PORT,
-            password=settings.REDIS_PASSWORD,
-            db=settings.REDIS_DB,
+            host=config["host"],
+            port=config["port"],
+            password=config["password"],
+            db=config["db"],
             socket_keepalive=True,
         )
 
 
-redis_client: RedisClientType = _initialize_redis_cluster()
+_default_redis_client: RedisClientType = _initialize_redis_cluster(
+    {
+        "use_redis_cluster": settings.USE_REDIS_CLUSTER,
+        "cluster_startup_nodes": settings.REDIS_CLUSTER_STARTUP_NODES,
+        "host": settings.REDIS_HOST,
+        "port": settings.REDIS_PORT,
+        "password": settings.REDIS_PASSWORD,
+        "db": settings.REDIS_DB,
+        "reinitialize_steps": settings.REDIS_REINITIALIZE_STEPS,
+    }
+)
+
+
+def _initialize_specialized_redis_cluster(
+    config: settings.RedisClusterConfig | None,
+) -> RedisClientType:
+    if config is None:
+        return _default_redis_client
+
+    return _initialize_redis_cluster(config)
+
+
+class RedisClients(TypedDict):
+    cache: RedisClientType
+    rate_limiter: RedisClientType
+    subscription_store: RedisClientType
+    replacements_store: RedisClientType
+    misc: RedisClientType
+
+
+redis_clients: RedisClients = {
```

Comment:
I think we should probably the actual mapping behind a `get_redis_client(name: str)` or something

Reply:
done